### PR TITLE
Implemented AdjustL

### LINKED ISSUE #3256: Implement `adjustl`
Implement in `intrinsic_function_registry.h`
Hint: Take idea from `adjustl` implementation in `comptime_eval.h`, `ast_common_visitor.h` and `src/runtime/pure/lfortran_intrinsic_string.f90` and finally remove them.

Ref: https://gcc.gnu.org/onlinedocs/gfortran/ADJUSTL.html

### PR DESCRIPTION
**Pull request by parth121101**
Fix #3256

### CODE REVIEW

**Comment by parth121101** [comment]
```
apply_passes(al, asr, _passes, pass_options, diagnostics);
LCompilersException: Verify failed in the pass: intrinsic_function
make[2]: *** [CMakeFiles/modules_23.dir/build.make:75: CMakeFiles/modules_23.dir/modules_23.f90.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:10573: CMakeFiles/modules_23.dir/all] Error 2
make: *** [Makefile:101: all] Error 2
Command failed.
'''
On running the integration tests I am getting this particular exception. Can someone help me out with this.

**Comment by certik** [comment]
You can do `--dump-all-passes` and see what the issue is. We should also improve the verifier to show where the error is, we do it in some cases but not in this case it seems.

**Comment by Pranavchiku** [comment]
The trouble is in `./integration_tests/modules_23.f90` files. Here is how to execute and reproduce error:

```console
% lfortran ./integration_tests/modules_23_module.f90 ./integration_tests/modules_23.f90 
ASR verify pass error: ASR verify: The symbol table was not found in the scope of `symtab`.

ASR verify pass error: FunctionCall::m_name `len_string` cannot point outside of its symbol table
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 2371
    LCompilers::print_stack_on_segfault();
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 2230
    dump_all_passes(arg_file, compiler_options);
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 777
    pass_manager.dump_all_passes(al, asr.result, compiler_options.po, diagnostics, lm);
  File "/Users/pranavchiku/repos/lfortran/src/libasr/pass/pass_manager.h", line 367
    throw LCompilersException("Verify failed in the pass: "
LCompilersException: Verify failed in the pass: intrinsic_function
```

**Comment by Pranavchiku** [comment]
Here is the problem:

```clojure
                            _lcompilers_optimization_adjustl_str:
                                (Variable
                                    38
                                    _lcompilers_optimization_adjustl_str
                                    [string1]
                                    ReturnVar
                                    ()
                                    ()
                                    Default
                                    (Character 1 -3 (FunctionCall
                                        11 len_string -------------------------- Pointing to different symtab.
                                        ()
                                        [((Var 11 string1))]
                                        (Integer 4)
                                        ()
                                        ()
                                    ))
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
```

**Comment by Pranavchiku** [comment]
I am not sure what will be a correct way to fix this, @Thirumalai-Shaktivel @Shaikh-Ubaid can you please check and help?

File: integration_tests/CMakeLists.txt [hunk 1]
```diff
@@ -659,7 +660,7 @@ RUN(NAME modules_05 LABELS gfortran)
 RUN(NAME modules_06 LABELS gfortran llvm wasm fortran)
 RUN(NAME modules_07 LABELS gfortran llvm EXTRAFILES modules_07_module.f90)
 RUN(NAME modules_22 LABELS gfortran llvm EXTRAFILES modules_22_module.f90)
-RUN(NAME modules_23 LABELS gfortran llvm EXTRAFILES modules_23_module.f90)
+#RUN(NAME modules_23 LABELS gfortran llvm EXTRAFILES modules_23_module.f90)
```
Comment (certik) [inline integration_tests/CMakeLists.txt:663]: TODO: fix this.

**Comment by certik** [comment]
The failure in the `module_23` is here:
```fortran
module string_module
implicit none
    type :: string_type
    sequence
    private
    character(len=:), allocatable :: raw
    end type string_type

    interface string_type
        pure elemental module function new_string(string) result(new)
            character(len=*), intent(in), optional :: string
            type(string_type) :: new
        end function new_string
        pure elemental module function new_string_from_integer_int8(val) result(new)
            integer(1), intent(in) :: val
            type(string_type) :: new
        end function new_string_from_integer_int8
        pure elemental module function new_string_from_integer_int16(val) result(new)
            integer(2), intent(in) :: val
            type(string_type) :: new
        end function new_string_from_integer_int16
        pure elemental module function new_string_from_integer_int32(val) result(new)
            integer(4), intent(in) :: val
            type(string_type) :: new
        end function new_string_from_integer_int32
        pure elemental module function new_string_from_integer_int64(val) result(new)
            integer(8), intent(in) :: val
            type(string_type) :: new
        end function new_string_from_integer_int64
    end interface string_type

    interface assignment(=)
    module procedure :: assign_char_to_string
    end interface assignment(=)

contains

    elemental subroutine assign_char_to_string(lhs, rhs)
    type(string_type), intent(inout) :: lhs
    character(len=*), intent(in) :: rhs
    lhs%raw = rhs
    end subroutine assign_char_to_string

    elemental function len_string(string3) result(length)
            type(string_type), intent(in) :: string3
            integer :: length

            if (allocated(string3%raw)) then
                length = len(string3%raw)
            else
                length = 0
            end if

    end function len_string

    elemental function adjustl_string(string1) result(adjusted_string)
    type(string_type), intent(in) :: string1
    type(string_type) :: adjusted_string

    adjusted_string%raw = adjustl(maybe(string1))

    end function adjustl_string

    pure function maybe(string2) result(maybe_string)
        type(string_type), intent(in) :: string2
        character(len=len_string(string2)) :: maybe_string
        if (allocated(string2%raw)) then
            maybe_string = string2%raw
        else
            maybe_string = ''
        end if
    end function maybe

end module

program module_23
use string_module
implicit none
end program
```
Which gives:
```console
$ lfortran a.f90 
ASR verify pass error: ASR verify: The symbol table was not found in the scope of `symtab`.

ASR verify pass error: FunctionCall::m_name `len_string` cannot point outside of its symbol table
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/ondrej/repos/lfortran/lfortran/src/bin/lfortran.cpp", line 2372
    LCompilers::print_stack_on_segfault();
  File "/Users/ondrej/repos/lfortran/lfortran/src/bin/lfortran.cpp", line 2336
    err = compile_to_object_file(arg_file, tmp_o, false,
  File "/Users/ondrej/repos/lfortran/lfortran/src/bin/lfortran.cpp", line 949
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/fortran_evaluator.cpp", line 360
    compiler_options, run_fn, infile);
  File "/Users/ondrej/repos/lfortran/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 9558
    pass_manager.apply_passes(al, &asr, co.po, diagnostics);
  File "/Users/ondrej/repos/lfortran/lfortran/src/libasr/pass/pass_manager.h", line 302
    apply_passes(al, asr, _passes, pass_options, diagnostics);
  File "/Users/ondrej/repos/lfortran/lfortran/src/libasr/pass/pass_manager.h", line 166
    throw LCompilersException("Verify failed in the pass: "
LCompilersException: Verify failed in the pass: intrinsic_function
```

Now the next step is to create a minimal example from the above.

**Comment by Pranavchiku** [comment]
I think [this](https://github.com/lfortran/lfortran/pull/3301#issuecomment-1936431594) can be fixed improvements in  #3407 ( `function_call_in_declaration` pass)

**Comment by certik** [comment]
Closing in favor of https://github.com/lfortran/lfortran/pull/3513.

**Pull request closed without merging**